feat(2fa): option for disallowing new enrollments on 2FA interface

mdtro opened the pull request:
Introduce an option to disallow new enrollments for a particular 2FA interface. For example, if we want to start phasing out SMS, we can set the `sms.disallow-new-enrollment` option to `True`.

With this set to `True`:
- users already enrolled in SMS 2FA will still be able to authenticate using SMS 2FA
  - they will only be able to view or delete the SMS 2FA enrollment details
- users who were **not** already enrolled in SMS 2FA will not see the option to enroll in it

**Frontend changes need to be merged first**: https://github.com/getsentry/sentry/pull/36345

**Already enrolled in SMS 2FA**
<img width="717" alt="image" src="https://user-images.githubusercontent.com/20070360/177429508-7e17e7bb-f6d7-4d24-b7dd-7cf94512c88e.png">

**Not already enrolled in SMS 2FA**
<img width="795" alt="image" src="https://user-images.githubusercontent.com/20070360/177429582-52a3bd56-15bb-444a-a16b-fd6a480c4f21.png">

mdtro committed:
feat(2fa): add option to 2FA interfaces to allow for disallowing of new enrollments

mdtro committed:
remove dead code

mdtro committed:
fix boolean logic

dashed reviewed (approved):
This looks good to me 👍 

Would you be able to add some test cases in `tests/sentry/auth/authenticators/` to increase confidence that `disallow_new_enrollment` is working? 👀

mdtro committed:
tests sms disallow enrollment

mdtro committed:
ref(2fa): move `disallow_new_enrollment` to base class with tests

- move the `disallow_new_enrollment` to the base `AuthenticatorInterface` class as a property
  - for better ergonomics and coverage of 2FA interfaces
  - better testability
- tests for `disallow_new_enrollment` on the sms, totp, and u2f interfaces
